### PR TITLE
Modlog: enhance search

### PR DESCRIPTION
**Pull request by sparkychildcharlie**
Split and carried over from https://github.com/Zarel/Pokemon-Showdown/pull/3972

Please do not merge this until the aforementioned PR is merged first.

[18:33:36] #Zarel: that just means I merge the modlog entry pullreq first
[18:33:40] #Zarel: and you can note that in the pullreqs

ORIGINAL OP:

Expanding Modlog's capabilities:
---
- using ``,`` to allow searching for multiple strings in the same query (AND)
- using ``|`` to allow accepting any one of terms listed (OR)
- using ``*`` as a wildcard in USERID searches
- using ``!`` to exclude any search term (NOT) - must be placed at the start of that search term
- requiring the same symbol at the front and end of string searches, adding a new symbol `` ` `` for specifying an userid search
- userid and string searches can be mixed - ``"pika" | chu`` will search for ``\bchu\b`` or anything with the 4 letters ``pika`` side by side

This allows more complex searches like:
complex search
chu

Regex Details
---
- Regex for complex searches formatted as ``/^${stuff_to_include}(?:${stuff_to_exclude}.)*$/i``
    - Example: ``/^(?=.*(\ba[\W_]*d[\W_]*d[\W_]*e[\W_]*d\b|\bd[\W_]*e[\W_]*m[\W_]*o[\W_]*t[\W_]*e[\W_]*d\b))(?:(?!\bc[\W_]*a[\W_]*r[\W_]*d\b).)*$/i`` 
        - original search: ``added | demoted, !card`` 
        - contains the word ``added`` or ``demoted`` AND CANNOT include the word ``card``

Fixing errors:
---
- putting ``|`` in the search query can no longer affect the line count (in the past putting ``|`` into a modlog search adds an additional field in ``onMessageDownstream``, effectively pushing the originally intended values one "block" down the array)
<br /><br /><br />
<sub>(also heavy use of ``const`` for non-changing variables ^^)</sub>

### CODE REVIEW

**Comment by sparkychildcharlie** [comment]
For a branch with both the PRs together, please refer to https://github.com/sparkychildcharlie/Pokemon-Showdown/tree/modlog

File: chat-plugins/modlog.js [hunk 1]
```diff
@@ -386,9 +429,21 @@ exports.commands = {
...
+
+		this.sendReplyBox(
+			`<div class='chat'>` +
+			`/modlog [roomid], [search] - Searches the moderator log - defaults to the current room unless specified otherwise.<br />` +
+			`- If you set [roomid] as [all], it searches for [search] on all rooms' moderator logs.<br />` +
+			`- If you set [roomid] as [public], it searches for [search] in all public rooms' moderator logs, excluding battles. Requires: % @ * # & ~<br />` +
+			`- Put <code style=''>"</code> around any search term to get the exact string, with symbols and spacing included, in any part of the modlog entry.  By default, it will search for userids only if quotation marks are not used.<br />` +
+			`- Use <code>&&</code> include additional search terms.  Results will include lines with all the terms.<br />` +
+			`- Use <code>||</code> to include additional search terms. Results will include lines with any or all of the terms.<br />` +
+			`- Use <code>!</code> to exclude any search term.  Results will not include lines containing that term. <br />` +
```
Comment (panpawn) [inline chat-plugins/modlog.js]: You don't need a space between the period and the ``<br />``

**Comment by sparkychildcharlie** [comment]
bumpity bumpity bump now that the other monster of a PR is done!

**Comment by Zarel** [comment]
I'll wait until the other one has undergone some live testing.

**Comment by Zarel** [comment]
Okay, so, what still needs to be done here?

**Comment by sparkychildcharlie** [comment]
I just need to fix the merge conflicts, get you to run the script to
convert in the other PR, which i will do after my university assignments
are done.  Sorry, I've been really really busy with class stuff.

**Comment by sparkychildcharlie** [comment]
More or less lost interest/loyalty for PS

**Pull request closed without merging**